web/src/settings_bots.js

```diff
@@ -511,6 +523,103 @@ export function show_edit_bot_info_modal(user_id, from_user_info_popover) {
     });
 }
 
+export function show_generate_url_modal(api_key) {
+    const default_message = $t_html({defaultMessage: "Integration URL will appear here."});
+    let selected_integration = "";
+
+    const streams = stream_data.subscribed_streams();
+    const html_body = render_generate_integration_url_modal({default_message, streams});
+    dialog_widget.launch({
+        html_heading: $t_html({defaultMessage: "Generate URL for an integration"}),
+        html_body,
+        id: "generate-integration-url-modal",
+        html_submit_button: $t_html({defaultMessage: "Copy URL"}),
+        on_click() {},
+        close_on_submit: true,
+    });
+
+    const $integration_url = $("#generate-integration-url-modal .integration-url");
+    const $integration_name = $("#generate-integration-url-modal #integration-name");
+    const $topic_input = $("#generate-integration-url-modal .topic-input");
+    const $topic_name = $("#generate-integration-url-modal #topic-name");
+    const $stream_name = $("#generate-integration-url-modal #stream-name");
+    const $override_topic = $("#generate-integration-url-modal #override-topic");
+    const $override_topic_group = $("#generate-integration-url-modal .override-topic-group");
+
+    new ClipboardJS("#generate-integration-url-modal .dialog_submit_button", {
+        text() {
+            return $integration_url.text();
+        },
+    });
+
+    $integration_name
+        .typeahead({
+            items: 5,
+            fixed: true,
+            source: () => page_params.realm_incoming_webhook_bots.map((bot) => bot.display_name),
+            updater(item) {
+                selected_integration = page_params.realm_incoming_webhook_bots.find(
+                    (bot) => bot.display_name === item,
+                ).name;
+                update_url();
+                return item;
+            },
+        })
+        .on("input", function () {
+            const current_value = $(this).val();
+            if (current_value === "") {
+                selected_integration = "";
+                update_url();
+            }
+        });
+
+    $stream_name.on("change", function () {
+        const stream_name = $(this).val();
+        if (stream_name === "") {
+            $override_topic.prop("checked", false).prop("disabled", true);
+            $override_topic_group.addClass("control-label-disabled");
+            $topic_input.addClass("hide");
+            $topic_name.val("");
+        } else {
+            $override_topic.prop("disabled", false);
+            $override_topic_group.removeClass("control-label-disabled");
+        }
+    });
+
+    $override_topic.on("change", function () {
+        const checked = $(this).prop("checked");
+        $topic_input.toggleClass("hide", !checked);
+        $topic_name.val("");
+        update_url();
```

Comment:
Since you are calling `update_url` for the `.update-url` handler, can we remove it from other places and just add that class to inputs which affect the url. I am not sure we would still need to call it from `updater` function but you can check.

Reply:
That is possible and would certainly make the code look cleaner.